File: app/src/main/java/com/nguyendevs/stkh/Manager/TranslateManager.java
package com.nguyendevs.stkh.Manager;

import android.app.AlertDialog;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.widget.EditText;
import android.widget.Toast;
import com.google.mlkit.nl.languageid.LanguageIdentification;
import com.google.mlkit.nl.languageid.LanguageIdentifier;
import com.google.mlkit.nl.translate.Translation;
import com.google.mlkit.nl.translate.Translator;
import com.google.mlkit.nl.translate.TranslatorOptions;
import com.nguyendevs.stkh.MainActivity;

public class TranslateManager {
    private MainActivity activity;
    private EditText txtResult;
    private Translator translator;
    private String originalTextSnapshot = null;
    public TranslateManager(MainActivity activity, EditText txtResult) {
        this.activity = activity;
        this.txtResult = txtResult; }
    public void showTargetLanguagePopup() {
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setTitle("Dịch sang ngôn ngữ nào?");
        String[] languages = {
                "Tiếng Việt",
                "Tiếng Anh",
                "Tiếng Nhật",
                "Tiếng Nga",
                "Tiếng Pháp",
                "Tiếng Tây Ban Nha",
                "Tiếng Đức",
                "Tiếng Trung",
                "Tiếng Hàn",
                "Tiếng Ý" };
        String[] languageCodes = {
                "Vietnamese",
                "English",
                "Japanese",
                "Russian",
                "French",
                "Spanish",
                "German",
                "Chinese",
                "Korean",
                "Italian" };
        builder.setItems(languages, (dialog, which) -> {
            String currentText = txtResult.getText().toString().trim();
            if (currentText.isEmpty()) {
                Toast.makeText(activity, "Không có nội dung để dịch.", Toast.LENGTH_SHORT).show();
                return; }
            if (originalTextSnapshot == null) {
                originalTextSnapshot = currentText; }
            String targetLang = languageCodes[which];
            detectAndTranslate(originalTextSnapshot, targetLang); });
        builder.show();  }
    public void resetOriginalTextSnapshot() {
        originalTextSnapshot = null; }
    private boolean isNetworkAvailable() {
        ConnectivityManager cm = (ConnectivityManager) activity.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = cm.getActiveNetworkInfo();
        return netInfo != null && netInfo.isConnected(); }
    private String convertToMLKitCode(String lang) {
        switch (lang) {
            case "Vietnamese": return com.google.mlkit.nl.translate.TranslateLanguage.VIETNAMESE;
            case "Japanese": return com.google.mlkit.nl.translate.TranslateLanguage.JAPANESE;
            case "Korean": return com.google.mlkit.nl.translate.TranslateLanguage.KOREAN;
            case "English": return com.google.mlkit.nl.translate.TranslateLanguage.ENGLISH;
            case "Russian": return com.google.mlkit.nl.translate.TranslateLanguage.RUSSIAN;
            case "Italian": return com.google.mlkit.nl.translate.TranslateLanguage.ITALIAN;
            case "Spanish": return com.google.mlkit.nl.translate.TranslateLanguage.SPANISH;
            case "French": return com.google.mlkit.nl.translate.TranslateLanguage.FRENCH;
            case "German": return com.google.mlkit.nl.translate.TranslateLanguage.GERMAN;
            case "Chinese": return com.google.mlkit.nl.translate.TranslateLanguage.CHINESE;
            default: return com.google.mlkit.nl.translate.TranslateLanguage.VIETNAMESE; } }
    private String getDisplayLanguageName(String lang) {
        switch (lang) {
            case "Vietnamese": return "Tiếng Việt";
            case "Japanese": return "Tiếng Nhật";
            case "Korean": return "Tiếng Hàn";
            case "English": return "Tiếng Anh";
            case "French": return "Tiếng Pháp";
            case "German": return "Tiếng Đức";
            case "Chinese": return "Tiếng Trung";
            case "Russian": return "Tiếng Nga";
            case "Spanish": return "Tiếng Tây Ban Nha";
            case "Italian": return "Tiếng Ý";
            default: return "Tiếng Việt"; } }
    private void detectAndTranslate(String text, String targetLang) {
        LanguageIdentifier identifier = LanguageIdentification.getClient();
        identifier.identifyLanguage(text)
                .addOnSuccessListener(languageCode -> {
                    if (languageCode.equals("unknown")) {
                        Toast.makeText(activity, "Không thể xác định ngôn ngữ.", Toast.LENGTH_SHORT).show();
                        return; }
                    translateText(text, languageCode, convertToMLKitCode(targetLang), targetLang);  })
                .addOnFailureListener(e -> {
                    Toast.makeText(activity, "Lỗi xác định ngôn ngữ: " + e.getMessage(), Toast.LENGTH_SHORT).show(); }); }
    private void translateText(String text, String sourceLang, String targetLangCode, String targetLangName) {
        TranslatorOptions options = new TranslatorOptions.Builder()
                .setSourceLanguage(sourceLang)
                .setTargetLanguage(targetLangCode)
                .build();
        translator = Translation.getClient(options);
        if (!isNetworkAvailable()) {
            Toast.makeText(activity, "Không có kết nối mạng. Đang thử dịch ngoại tuyến.", Toast.LENGTH_SHORT).show(); }
        translator.downloadModelIfNeeded()
                .addOnSuccessListener(unused -> {
                    translator.translate(text)
                            .addOnSuccessListener(translatedText -> {
                                txtResult.setText(translatedText);
                                TextToSpeechManager tts = activity.getTextToSpeechManager();
                                tts.setDetectedLanguage(targetLangName);
                                tts.speakText(translatedText); })
                            .addOnFailureListener(e -> {
                                Toast.makeText(activity, "Dịch thất bại: " + e.getMessage(), Toast.LENGTH_LONG).show(); }); })
                .addOnFailureListener(e -> {
                    Toast.makeText(activity, "Không thể tải mô hình dịch: " + e.getMessage(), Toast.LENGTH_LONG).show(); }); }
    public void destroy() {
        if (translator != null) {
            translator.close(); } } }

File: app/src/main/java/com/nguyendevs/stkh/manager/TranslateManager.kt
package com.nguyendevs.stkh.manager

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.widget.EditText
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.mlkit.nl.languageid.LanguageIdentification
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.Translator
import com.google.mlkit.nl.translate.TranslatorOptions
import com.google.mlkit.nl.translate.TranslateLanguage
import com.nguyendevs.stkh.util.showToast

class TranslateManager(
    private val context: Context,
    private val txtResult: EditText
) : ITranslator {

    private var translator: Translator? = null
    private var originalTextSnapshot: String? = null

    var onTranslated: ((text: String, targetLang: String) -> Unit)? = null

    companion object {
        private val languageDisplayMap: Map<String, String> = mapOf(
            "Vietnamese" to "Tiếng Việt",
            "English" to "Tiếng Anh",
            "Japanese" to "Tiếng Nhật",
            "Russian" to "Tiếng Nga",
            "French" to "Tiếng Pháp",
            "Spanish" to "Tiếng Tây Ban Nha",
            "German" to "Tiếng Đức",
            "Chinese" to "Tiếng Trung",
            "Korean" to "Tiếng Hàn",
            "Italian" to "Tiếng Ý"
        )

        private val languageToMLKitMap: Map<String, String> = mapOf(
            "Vietnamese" to TranslateLanguage.VIETNAMESE,
            "English" to TranslateLanguage.ENGLISH,
            "Japanese" to TranslateLanguage.JAPANESE,
            "Russian" to TranslateLanguage.RUSSIAN,
            "French" to TranslateLanguage.FRENCH,
            "Spanish" to TranslateLanguage.SPANISH,
            "German" to TranslateLanguage.GERMAN,
            "Chinese" to TranslateLanguage.CHINESE,
            "Korean" to TranslateLanguage.KOREAN,
            "Italian" to TranslateLanguage.ITALIAN
        )
    }

    override fun showTargetLanguagePopup() {
        val languages = languageDisplayMap.keys.toTypedArray()
        val displayNames = languageDisplayMap.values.toTypedArray()

        MaterialAlertDialogBuilder(context)
            .setTitle("Dịch sang ngôn ngữ nào?")
            .setItems(displayNames) { _, which ->
                val currentText = txtResult.text.toString().trim()
                if (currentText.isEmpty()) {
                    context.showToast("Không có nội dung để dịch.")
                    return@setItems
                }
                if (originalTextSnapshot == null) {
                    originalTextSnapshot = currentText
                }
                detectAndTranslate(originalTextSnapshot!!, languages[which])
            }
            .show()
    }

    override fun resetOriginalTextSnapshot() {
        originalTextSnapshot = null
    }

    private fun isNetworkAvailable(): Boolean {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val network = cm.activeNetwork ?: return false
            val caps = cm.getNetworkCapabilities(network) ?: return false
            caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
        } else {
            @Suppress("DEPRECATION")
            cm.activeNetworkInfo?.isConnected == true
        }
    }

    private fun detectAndTranslate(text: String, targetLang: String) {
        val identifier = LanguageIdentification.getClient()
        identifier.identifyLanguage(text)
            .addOnSuccessListener { languageCode ->
                if (languageCode == "und") {
                    context.showToast("Không thể xác định ngôn ngữ.")
                    return@addOnSuccessListener
                }
                val targetMLKitCode = languageToMLKitMap[targetLang]
                    ?: TranslateLanguage.VIETNAMESE
                translateText(text, languageCode, targetMLKitCode, targetLang)
            }
            .addOnFailureListener { e ->
                context.showToast("Lỗi xác định ngôn ngữ: ${e.message}")
            }
    }

    private fun translateText(
        text: String,
        sourceLang: String,
        targetLangCode: String,
        targetLangName: String
    ) {
        val options = TranslatorOptions.Builder()
            .setSourceLanguage(sourceLang)
            .setTargetLanguage(targetLangCode)
            .build()

        translator = Translation.getClient(options)

        if (!isNetworkAvailable()) {
            context.showToast("Không có kết nối mạng. Đang thử dịch ngoại tuyến.")
        }

        translator!!.downloadModelIfNeeded()
            .addOnSuccessListener {
                translator!!.translate(text)
                    .addOnSuccessListener { translatedText ->
                        txtResult.setText(translatedText)
                        onTranslated?.invoke(translatedText, targetLangName)
                    }
                    .addOnFailureListener { e ->
                        context.showToast("Dịch thất bại: ${e.message}", android.widget.Toast.LENGTH_LONG)
                    }
            }
            .addOnFailureListener { e ->
                context.showToast("Không thể tải mô hình dịch: ${e.message}", android.widget.Toast.LENGTH_LONG)
            }
    }

    override fun destroy() {
        translator?.close()
        translator = null
    }
}
